src/suites/cts/validation/error_scope.spec.ts

```diff
@@ -0,0 +1,106 @@
+export const description = `
+error scope validation tests.
+`;
+
+import { getGPU } from '../../../framework/gpu/implementation.js';
+import { Fixture, TestGroup } from '../../../framework/index.js';
+
+class F extends Fixture {
+  device: GPUDevice = undefined!;
+
+  async init(): Promise<void> {
+    super.init();
+    const gpu = getGPU();
+    const adapter = await gpu.requestAdapter();
+    this.device = await adapter.requestDevice();
+  }
+
+  createErrorBuffer(): void {
+    this.device.createBuffer({
+      size: 1024,
+      usage: 0xffff, // Invalid GPUBufferUsage
+    });
+  }
+
+  async expectUncapturedError(fn: Function): Promise<void> {
+    return this.asyncExpectation(async () => {
+      const promise = new Promise(resolve => {
+        this.device.addEventListener('uncapturederror', resolve, { once: true });
+      });
+
+      fn();
+
+      await promise;
+    });
+  }
+}
+
+export const g = new TestGroup(F);
+
+g.test('simple case where the error scope catches an error', async t => {
+  t.device.pushErrorScope('validation');
+
+  t.createErrorBuffer();
+
+  const error = await t.device.popErrorScope();
+  t.expect(error instanceof GPUValidationError);
+});
+
+g.test('errors bubble to the parent scope if not handled by the current scope', async t => {
+  t.device.pushErrorScope('validation');
+  t.device.pushErrorScope('out-of-memory');
+
+  t.createErrorBuffer();
+
+  {
+    const error = await t.device.popErrorScope();
+    t.expect(error === null);
+  }
+  {
+    const error = await t.device.popErrorScope();
+    t.expect(error instanceof GPUValidationError);
+  }
+});
+
+g.test('if an error scope matches an error it does not bubble to the parent scope', async t => {
+  t.device.pushErrorScope('out-of-memory');
+  t.device.pushErrorScope('validation');
+
+  t.createErrorBuffer();
+
+  {
+    const error = await t.device.popErrorScope();
+    t.expect(error instanceof GPUValidationError);
+  }
+  {
+    const error = await t.device.popErrorScope();
+    t.expect(error === null);
+  }
+});
+
+g.test('if no error scope handles an error it fires an uncapturederror event', async t => {
+  t.device.pushErrorScope('out-of-memory');
+
+  t.expectUncapturedError(() => {
```

Comment:
This should require an await. And the `asyncExpectation` should be detecting the fact that we don't await it. Is it not catching this?

Reply:
It does require an await. Thanks!
`asyncExpectation` was not catching this error. How could it?

Reply:
The purpose of asyncExpectation is that it tracks the numOutstandingAsyncExpectations, which doesn't decrement until after the asynchronous expectation function finishes. So if the test function (`async t => ...`) finishes before numOutstandingAsyncExpectations goes back down to 0, it's supposed to throw an error.

That said, Corentin asked that asyncExpectation change so that it just adds promises that the harness will wait on automatically at the end of the test, instead of asserting. That may be nicer.

Reply:
Actually, waiting on all async expectations at the end is not ideal for this particular case. We want to make sure that the uncapturederror is coming from the right place - so we want to make sure that we the uncapturederror expectation to pass or fail before doing any more work.

Technically, it would be best if we unregistered the uncapturederror handler if we reject on timeout, but omitting that can't cause false positives so it's OK.